modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -48,5 +59,67 @@ object ConstructionService {
 
       NonEmptyList.fromList(accountIdentifiers)
     }
+
+    def transactionToOperations(transaction: Transaction): Either[String, NonEmptyList[Operation]] =
+      refineV[Not[Equal[0L]]](transaction.amount.value.value).flatMap { amount =>
+        val operations: Either[String, List[Operation]] = refineV[Not[Equal[0L]]](-1L * transaction.amount.value.value).map { negAmount =>
+          val negativeOperation = Operation(
+            OperationIdentifier(OperationIndex(0L)),
+            none,
+            OperationType.Transfer,
+            none,
+            AccountIdentifier(transaction.destination, none).some,
+            Amount(AmountValue(negAmount), DAG).some
+          )
+
+          val positiveOperation = Operation(
+            OperationIdentifier(OperationIndex(1L)),
+            none,
+            OperationType.Transfer,
+            none,
+            AccountIdentifier(transaction.destination, none).some,
+            Amount(AmountValue(amount), DAG).some
+          )
+
+          List(negativeOperation, positiveOperation)
+        }
+
+        operations.flatMap { operationList =>
+          NonEmptyList.fromList(operationList) match {
+            case Some(value) => value.asRight
+            case None        => "No valid operations found.".asLeft
+          }
+        }
+      }
+
+    def parseSignedTransaction(hex: Hex): F[Either[ConstructionError, ConstructionParse.ServiceReturn]] =
+      KryoSerializer[F]
+        .deserialize[Signed[Transaction]](hex.toBytes)
+        .traverse { signedTransaction =>
+          transactionToOperations(signedTransaction).traverse { operationsList =>
+            signedTransaction.proofs.toNonEmptyList.traverse(signature => signature.id.hex.toPublicKey).map { proofs =>
+              val proofAccountIdentifiers = proofs.map(_.toAddress).map { address =>
+                AccountIdentifier(address, none)
+              }
+
+              ConstructionParse.ServiceReturn(operationsList, proofAccountIdentifiers.some)
+            }
+          }.map(_.left.map(_ => MalformedTransaction))
+        }
+        .map(_.left.map(_ => MalformedTransaction).flatten)
+
+    def parseUnsignedTransaction(hex: Hex): Either[ConstructionError, ConstructionParse.ServiceReturn] =
+      KryoSerializer[F]
+        .deserialize[Transaction](hex.toBytes)
+        .flatMap(transactionToOperations)
+        .map(operations => ConstructionParse.ServiceReturn(operations, none))
+        .leftMap(_ => MalformedTransaction)
+
+    def parseTransaction(hex: Hex, isSigned: Boolean): EitherT[F, ConstructionError, ConstructionParse.ServiceReturn] =
+      EitherT(if (isSigned) {
```

Comment:
You can just lift `Either[ConstructionError, ConstructionParse.ServiceReturn]` from `parseUnsignedTransaction` to `EitherT[F, ConstructionError, ConstructionParse.ServiceReturn]` https://typelevel.org/cats/datatypes/eithert.html#from-either-a-b-or-f-either-a-b-to-eithert-f-a-b

```scala
import cats.syntax.either._

if (isSigned) 
  parseSignedTransaction(hex)
else
  parseUnsignedTransaction.toEitherT

```

Reply:
Changed.